### ISSUE TITLE
Feature request: alert/notifications should send OK notification when state recovers

### ISSUE THREAD

**Issue by sdanieru** [rancher/rancher#18236]
**Feature request**

**Steps to reproduce (least amount of steps as possible):**
1. Setup an alert for something that represents a state you can easily change for testing (e.g. node cpu threshold to 1%)
2. connect it to a notifier that you can easily monitor alert signaling on (e.g. a slack channel)
3. Note in the Rancher UI when the alert state changes to 'Alerting'
4. Alert is sent to respective notifier after specified time period in alert.
5. change threshold in alert back to a sane value, like 95% or whatever
6. Observe alert status in Rancher UI changes back to green 'Active'

**Result:**
No 'OK' or 'resolved' notification is sent to the notifier when the status changes back to healthy.
This means, for example, if I have an alert sent to PagerDuty and then state recovers in Rancher, PagerDuty is never informed of the recovery and still shows the alert is active.
We have several other Alert sources that properly signal when a state is healthy again (e.g. pingdom, grafana, etc.); it would be great if Rancher did the same.

**Expected Result**
Rancher should send a cleared/OK/resolved/healthy signal to all the Notifiers when the alert goes back to a healthy state.

**Environment information**
- Rancher version: 2.1.6
- Installation option: single

**Comment by sdanieru** [comment]
Is there any chance of getting this in? For us it would really improve the alert/notification usability of Rancher.

**Comment by sdanieru** [comment]
Another plug for this feature request - it would make the alert/notifications in Rancher more useful!

**Comment by loganhz** [comment]
Thanks for the report. We will look at it.

**Comment by ankur111** [comment]
I am a bit unclear on the expectation. 

System encounters error - Alerts are triggered
We poll it periodically and keep sending alerts. 
Once the system is back to normal, alerts stop (but polling continues in the background)

If we start sending "all clear" alerts, we now have to determine when the last all clear was sent, has the system gone down since then, was an all clear alert already sent (and if it wasn't received then this whole thing breaks down), etc.

In short, once the error alerts stop going out, isn't that a clearer indication that everything is back to normal?

**Comment by sdanieru** [comment]
As near as I can tell, everything I've integrated with PagerDuty, excepting Rancher, follows a 'push state change' webhook model (e.g. https://www.pingdom.com/resources/webhooks/), and this is what PagerDuty expects. That's why I'm ending up with alerts in PagerDuty that, even after Rancher has stopped sending alerts, always need to be manually resolved.
I don't know if there is a way to tell PagerDuty to expect the polling model you mention, where the absence of alerts indicates the alert has cleared - I haven't looked into it, but given that the send interval is variable and could be configured for 1m or 5h or not at all in Rancher, this approach doesn't seem wise as the receiver has no way of knowing what this send interval might be and when it should assume 'another alert hasn't come' and clear the alert. 

Also, I don't follow your concerns about the 'push state change' model: if Rancher is able to successfully push, for example, a 'clear' state change over TCP/HTTP for a given alert ID to registered recipient(s), why would it need to track when the last all clear was sent or worry about the message not having been received? I would think at that point it can consider the communication a successful hand-off: job done and forget about that given alert ID.

As I see it, the push state change model is simple and elegant:
*when a new issue is detected, create an alert w/ ID, and notify recipient(s) of new status
*when status changes, send msg to recipient(s) with alert ID and new status
*When a 'clear' msg is successfully sent for a given alert ID, forget it.

With this model after a 'clear' is sent and your local polling indicates that the system is down again, it simply creates a new alert ID and sends it out w/ status, which makes sense since it could reasonably be considered a new alert. Also, if a recipient wants to keep receiving messages at a given interval as long as an alert is active (might be useful on slack for examples), Rancher already has the functionality for that under 'Advanced options'.

**Comment by loganhz** [comment]
@sdanieru  is it the same issue with https://github.com/rancher/rancher/issues/17962 ?

**Comment by sdanieru** [comment]
thanks @loganhz - that looks to be what I'm after, at least in part. I'll make further comment there.

**Comment by sdanieru** [comment]
Looking a little closer, it's slightly different. In [#17962](https://github.com/rancher/rancher/issues/17962), all the recipients mentioned by the OP don't need to correlate alert messages to a given alert (email or slack), whereas in PagerDuty alert messages do need to be correlated so it can de-dup duplicate messages, and properly ack or resolve an existing alert.

In particular, PagerDuty needs a unique id to associate with the alert; in PagerDuty API v1 it's called the 'incident_key', and 'dedup_key' in v2. In either case when a new alert is sent, the originator can supply the value, or if it's not supplied, PagerDuty will generate and return a unique alert value in the response; this value is then used to correlate all future messages pertaining to the given alert.

PagerDuty refs:
[API v1](https://v2.developer.pagerduty.com/docs/events-api)
[API v2](https://v2.developer.pagerduty.com/docs/events-api-v2)

**Comment by sdanieru** [comment]
So in short, this feature request has morphed into a 'please support alert integration with PagerDuty'. If such functionality was in place, it would play along nicely with Rancher's current model of 'keep sending alerts until the issue is resolved' since PagerDuty would de-dup, ack or resolve based on the unique alert ID.

**Comment by stefanlasiewski** [comment]
> @stefanlasiewski I agree, sending out a 'resolved' message is a better model than assuming silence means all is back to normal.

I saw this message yesterday in my Slack channel for alerts, for example. I would feel better if I saw an 'All clear' message indicating that etcd had actually recovered. If I don't see that, I need to log in to confirm that everything is okay:

![image](https://user-images.githubusercontent.com/142752/63984356-c36f0100-ca7f-11e9-9beb-7261f1d2c74f.png)

**Comment by axeal** [comment]
send_resolved appears to be a receiver parameter in alertmanager, so exposing this in the Rancher notifer options would enable users to configure whether to send resolved notifications (https://prometheus.io/docs/alerting/configuration/)

**Comment by axeal** [comment]
Support for configuring send resolved on notifiers was added in Rancher v2.3.0:
https://github.com/rancher/rancher/issues/17962
https://github.com/rancher/ui/pull/3294
